### PR TITLE
Add distributed moving laser heating

### PR DESCRIPTION
**Pull request by MarkMa1990**
This program aims to solve the non-uniform isotropic heat equation with distributed memory.

as suggested by wolfgang, the program is moved here.
https://github.com/dealii/dealii/pull/13023#issuecomment-984792876

### CODE REVIEW

File: Distributed_Moving_Laser_Heating/Distributed_Moving_Laser_Heating.cc [hunk 1]
```diff
@@ -0,0 +1,716 @@
...
+
+// the following files are for petsc, if you want to use
+// pesct instead of trilinos, you can uncomment the following
+
+/*
+#include <deal.II/lac/petsc_parallel_vector.h>
+#include <deal.II/lac/petsc_parallel_sparse_matrix.h>
+#include <deal.II/lac/petsc_solver.h>
+#include <deal.II/lac/petsc_precondition.h>
+*/
```
Comment (peterrum) [inline Distributed_Moving_Laser_Heating/Distributed_Moving_Laser_Heating.cc:48]: ?

File: Distributed_Moving_Laser_Heating/Distributed_Moving_Laser_Heating.cc [hunk 2]
```diff
@@ -0,0 +1,716 @@
...
+#include <deal.II/distributed/tria.h>
+
+#include <deal.II/numerics/data_out.h>
+#include <fstream>
+#include <iostream>
+
+
+#include <deal.II/base/logstream.h>
+// this library is deprecated. use affine_constraints.h instead
+//#include <deal.II/lac/constraint_matrix.h>
```
Comment (peterrum) [inline Distributed_Moving_Laser_Heating/Distributed_Moving_Laser_Heating.cc:72]:
```suggestion
```
```suggestion
// this library is deprecated. use affine_constraints.h instead
//#include <deal.II/lac/constraint_matrix.h>
```

Comment (peterrum) [inline Distributed_Moving_Laser_Heating/Distributed_Moving_Laser_Heating.cc:72]:
```suggestion
```
```suggestion
// this library is deprecated. use affine_constraints.h instead
//#include <deal.II/lac/constraint_matrix.h>
```

File: Distributed_Moving_Laser_Heating/Distributed_Moving_Laser_Heating.cc [hunk 3]
```diff
@@ -0,0 +1,716 @@
...
+// laser parameters, and heat characteristics of materials involved. Boundary
+// and initial conditions are defined in boundaryInit.h. The rightHandSide.h
+// defines the heat source, which in this work is a moving Gaussian beam.
+
+#ifndef GLOBAL_PARA
+#define GLOBAL_PARA
+#include "./globalPara.h"
+#include "./boundaryInit.h"
+#include "./rightHandSide.h"
+#endif
```
Comment (peterrum) [inline Distributed_Moving_Laser_Heating/Distributed_Moving_Laser_Heating.cc:97]: Is  this ifdef needed?

Comment (peterrum) [inline Distributed_Moving_Laser_Heating/Distributed_Moving_Laser_Heating.cc:97]: Is  this ifdef needed?

File: Distributed_Moving_Laser_Heating/Distributed_Moving_Laser_Heating.cc [hunk 4]
```diff
@@ -0,0 +1,716 @@
...
+
+  constraints_T.clear();
+  constraints_T.reinit (locally_relevant_dofs);
+  DoFTools::make_hanging_node_constraints (dof_handler, constraints_T);
+/*
+  VectorTools::interpolate_boundary_values (dof_handler,
+                                            BOUNDARY_NUM,
+                                            BoundaryValues<dim>(),
+                                            constraints_T);                                      
+*/
```
Comment (peterrum) [inline Distributed_Moving_Laser_Heating/Distributed_Moving_Laser_Heating.cc:242]:
```suggestion
```

Comment (peterrum) [inline Distributed_Moving_Laser_Heating/Distributed_Moving_Laser_Heating.cc:242]:
```suggestion
```

File: Distributed_Moving_Laser_Heating/Distributed_Moving_Laser_Heating.cc [hunk 5]
```diff
@@ -0,0 +1,716 @@
...
+  left_system_matrix_T = 0;
+  right_system_matrix_T = 0;
+
+  system_rhs_T = 0;
+
+  typename DoFHandler<dim>::active_cell_iterator
+  cell = dof_handler.begin_active(),
+  endc = dof_handler.end();
+
+  for (; cell!=endc; ++cell)
```
Comment (peterrum) [inline Distributed_Moving_Laser_Heating/Distributed_Moving_Laser_Heating.cc:304]:
```suggestion
  for (const auto cell : dof_handler.active_cell_iterator())
```

Comment (peterrum) [inline Distributed_Moving_Laser_Heating/Distributed_Moving_Laser_Heating.cc:304]:
```suggestion
  for (const auto cell : dof_handler.active_cell_iterator())
```

File: Distributed_Moving_Laser_Heating/Distributed_Moving_Laser_Heating.cc [hunk 6]
```diff
@@ -0,0 +1,716 @@
...
+
+   data_out.build_patches ();
+
+   const std::string filename = ("solution-" +
+                                Utilities::int_to_string (output_num, step_N) +
+                                "." +
+                                Utilities::int_to_string (triangulation.locally_owned_subdomain(),4) +
+                                ".vtu");
+   std::ofstream output (filename.c_str());
+   data_out.write_vtu (output);
```
Comment (peterrum) [inline Distributed_Moving_Laser_Heating/Distributed_Moving_Laser_Heating.cc:567]: You could use `write_vtu_in_parallel()` or `write_vtu_with_pvtu_record()` instead.

Comment (peterrum) [inline Distributed_Moving_Laser_Heating/Distributed_Moving_Laser_Heating.cc:567]: You could use `write_vtu_in_parallel()` or `write_vtu_with_pvtu_record()` instead.

File: Distributed_Moving_Laser_Heating/Distributed_Moving_Laser_Heating.cc [hunk 7]
```diff
@@ -0,0 +1,716 @@
...
+                                            BOUNDARY_NUM,
+                                            BoundaryValues<dim>(),
+                                            constraints_T);                                      
+*/
+  constraints_T.close();
+
+  DynamicSparsityPattern dsp_T(locally_relevant_dofs);
+  DoFTools::make_sparsity_pattern (dof_handler, dsp_T,constraints_T,false);
+  SparsityTools::distribute_sparsity_pattern (dsp_T,
+                                              dof_handler.n_locally_owned_dofs_per_processor(),
```
Comment (peterrum) [inline Distributed_Moving_Laser_Heating/Distributed_Moving_Laser_Heating.cc:248]: This does not compile with master. Could you check out what the issue is?

Comment (peterrum) [inline Distributed_Moving_Laser_Heating/Distributed_Moving_Laser_Heating.cc:248]: This does not compile with master. Could you check out what the issue is?

File: Distributed_Moving_Laser_Heating/rightHandSide.h [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+};
+
+//#----------------------------------------------------------
+//# Implementation
+//
+template <int dim>
+double RightHandside<dim>::value_v2 (const Point<dim> &p)
+{
+
+    double Rnp_inter;
```
Comment (peterrum) [inline Distributed_Moving_Laser_Heating/rightHandSide.h:31]: this variable is not used/

Comment (peterrum) [inline Distributed_Moving_Laser_Heating/rightHandSide.h:31]: this variable is not used/

File: Distributed_Moving_Laser_Heating/doc/entry-name [hunk 9]
```diff
@@ -0,0 +1 @@
+Laplace equation coupled to an external simulation program
```
Comment (peterrum) [inline Distributed_Moving_Laser_Heating/doc/entry-name:1]: ?

Comment (peterrum) [inline Distributed_Moving_Laser_Heating/doc/entry-name:1]: ?

File: Distributed_Moving_Laser_Heating/doc/tooltip [hunk 10]
```diff
@@ -0,0 +1 @@
+Laplace equation surface coupled to an external simulation program (here simply a fancy boundary condition) using the coupling library preCICE.
```
Comment (peterrum) [inline Distributed_Moving_Laser_Heating/doc/tooltip:1]: ?

Comment (peterrum) [inline Distributed_Moving_Laser_Heating/doc/tooltip:1]: ?

File: Distributed_Moving_Laser_Heating/docs_old/intro.dox [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+@f{align*}
+M_{i,j} &= \int \rho C_m T^{'}_i T^{'}_j dV\\
+A_{i,j} &= \int k \nabla T^{'}_i \nabla T^{'}_j dV \\
+F^n_i &= \int T^{'}_i f^n(\mathbf x) dV
+@f}
+
+where $n$ is the time step number, and $dt$ is the time step.
+
+<h3> The initial condition </h3>
+Deall.II has provide an interpolation function (VectorTools::interpolate) to distribute initial conditions. However, in this tutorial,
```
Comment (peterrum) [inline Distributed_Moving_Laser_Heating/docs_old/intro.dox:30]:
```suggestion
deall.II provides an interpolation function (VectorTools::interpolate) to distribute initial conditions. However, in this tutorial,
```
```suggestion
Deall.II has provide an interpolation function (VectorTools::interpolate) to distribute initial conditions. However, in this tutorial,
```

Comment (peterrum) [inline Distributed_Moving_Laser_Heating/docs_old/intro.dox:30]:
```suggestion
deall.II provides an interpolation function (VectorTools::interpolate) to distribute initial conditions. However, in this tutorial,
```
```suggestion
Deall.II has provide an interpolation function (VectorTools::interpolate) to distribute initial conditions. However, in this tutorial,
```

File: Distributed_Moving_Laser_Heating/README.md [hunk 12]
```diff
@@ -0,0 +1,183 @@
+
+# Introduction
+In this tutorial, the studied problem is to simulate temperature distributions of a sample under a moving laser. Light penetrates the substrate without loss. The top-covered thin-film is, however, a light absorbing material. For simplicity, the thin-film is assumed to be TiO$_2$ mixed with silver nanoparticles, which supports light heating of the material by absorbing light energy. For this tutorial, we only consider the isotropic absorption. Figure \ref{fgr:s1} illustrates the sketch of the problem. The absorption coefficient is assumed to be $10^4 m^{-1}$. The substrate is glass. The thickness of the thin-film is assumed to be 400 nm. The spot size at the top of thin film is $20 \mu m$ at $e^{-2}$. The writing speed is assumed to be 10 mm/s. The laser power is 0.4 W. The time step is set as 10 $\mu s$. The initial position of laser center is $-50 \mu m$ with 50 $\mu m$ distance away from the left boundary to avoid boundary effects.
+
+## Illustration of the problem
+
+![illustration](./images/structure-2d.png)
+
+## numerical results
```
Comment (peterrum) [inline Distributed_Moving_Laser_Heating/README.md:9]:
```suggestion
## Numerical results
```
```suggestion
## numerical results
```

Comment (peterrum) [inline Distributed_Moving_Laser_Heating/README.md:9]:
```suggestion
## Numerical results
```
```suggestion
## numerical results
```

**Review by peterrum — COMMENTED** [review]
Sorry for getting to this so later. I left some comments. Looks overall good.

File: Distributed_Moving_Laser_Heating/doc/entry-name (hunk 9 above)
Comment (bangerth) [reply Distributed_Moving_Laser_Heating/doc/entry-name:1]: I think that what @peterrum is wondering about is that that is probably not the title you wanted to give your program :-)

**Comment by bangerth** [comment]
Oh, did we never notice that you updated all of the things that @peterrum had commented on? In other words, is this ready to be merged from your side?

**Comment by bangerth** [comment]
My apologies that none of us saw that! In the future, add a comment whenever you address comments and we know that we should take another look.

**Review by marcfehling — APPROVED** [review]
I've made a few changes so that your program compiles and runs with the latest deal.II 9.4.0 release.

I have one question: What is the purpose of the `docs_old` folder? If it doesn't serve any purpose, I would vote to delete it.

Otherwise I would say that after @peterrum's extensive review this is ready to merge.

**Comment by marcfehling** [comment]
I did another force push to make sure that this code compiles on the new CI workflow introduced in #142.

**Review by peterrum — APPROVED** [review]
Thanks!

**Pull request merged**